portal-bridge/src/types/state/storage/evm_db.rs

```diff
@@ -0,0 +1,245 @@
+use std::{path::PathBuf, sync::Arc};
+
+use crate::types::state::storage::error::Error;
+use alloy_consensus::constants::KECCAK_EMPTY;
+use alloy_primitives::{Address, B256, U256};
+use alloy_rlp::{Decodable, EMPTY_STRING_CODE};
+use eth_trie::{EthTrie, Trie};
+use ethportal_api::types::state_trie::account_state::AccountState as AccountStateInfo;
+use parking_lot::Mutex;
+use revm::{DatabaseCommit, DatabaseRef};
+use revm_primitives::{keccak256, Account, AccountInfo, Bytecode, HashMap};
+use rocksdb::DB as RocksDB;
+
+use super::{
+    account::{Account as RocksAccount, AccountState as RocksAccountState},
+    account_db::AccountDB,
+    trie_db::TrieRocksDB,
+    utils::setup_rocksdb,
+};
+
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct EvmDB {
+    /// The underlying database.
+    pub db: Arc<RocksDB>,
+    /// To get proofs and to verify trie state.
+    pub trie: Arc<Mutex<EthTrie<TrieRocksDB>>>,
+}
+
+impl EvmDB {
+    pub fn new(path: Option<PathBuf>) -> anyhow::Result<Self> {
+        let db = Arc::new(setup_rocksdb(path)?);
+        db.put(KECCAK_EMPTY, Bytecode::new().bytes().as_ref())?;
+        db.put(B256::ZERO, Bytecode::new().bytes().as_ref())?;
+        let trie = Arc::new(Mutex::new(EthTrie::new(Arc::new(TrieRocksDB::new(
+            false,
+            db.clone(),
+        )))));
+        Ok(Self { db, trie })
+    }
+
+    /// Inserts the account's code into the cache.
+    ///
+    /// Accounts objects and code are stored separately in the cache, this will take the code from
+    /// the account and instead map it to the code hash.
+    ///
+    /// Note: This will not insert into the underlying external database.
+    pub fn insert_contract(&mut self, account: &mut AccountInfo) {
+        if let Some(code) = &account.code {
+            if !code.is_empty() {
+                if account.code_hash == KECCAK_EMPTY {
+                    account.code_hash = code.hash_slow();
+                }
+
+                self.db
+                    .put(account.code_hash, code.bytes().as_ref())
+                    .expect("Inserting contract shouldn't fail");
+            }
+        }
+        if account.code_hash == B256::ZERO {
+            account.code_hash = KECCAK_EMPTY;
+        }
+    }
+}
+
+impl DatabaseCommit for EvmDB {
+    fn commit(&mut self, changes: HashMap<Address, Account>) {
+        for (address, mut account) in changes {
```

Comment:
nit: The `keccak256(address)` is used many times below. Many calculate it once at the start of the loop